File: LLD/Design-Patterns/Observer/code/Observer.cpp
// Observer Pattern — a StockTicker (Subject) notifies registered Observers on price change.
// Compile: g++ -std=c++14 Observer.cpp -o observer
// Run:     ./observer

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class StockObserver {
public:
    virtual void update(const std::string& symbol, double price) = 0;
    virtual ~StockObserver() = default;
};

class StockTicker {
public:
    explicit StockTicker(std::string symbol) : symbol_(std::move(symbol)), price_(0.0) {}

    void attach(StockObserver* observer) {
        observers_.push_back(observer);
    }

    void detach(StockObserver* observer) {
        observers_.erase(std::remove(observers_.begin(), observers_.end(), observer), observers_.end());
    }

    void setPrice(double newPrice) {
        price_ = newPrice;
        notifyObservers();
    }

private:
    std::string symbol_;
    double price_;
    std::vector<StockObserver*> observers_;

    void notifyObservers() {
        for (auto* observer : observers_) {
            observer->update(symbol_, price_);
        }
    }
};

class MobileDisplay : public StockObserver {
public:
    void update(const std::string& symbol, double price) override {
        std::cout << "[MobileDisplay] " << symbol << " is now $" << price << std::endl;
    }
};

class EmailAlert : public StockObserver {
public:
    explicit EmailAlert(double threshold) : threshold_(threshold) {}

    void update(const std::string& symbol, double price) override {
        if (price >= threshold_) {
            std::cout << "[EmailAlert] " << symbol << " crossed $" << threshold_ << " — now $" << price << std::endl;
        }
    }

private:
    double threshold_;
};

int main() {
    StockTicker ticker("ACME");

    MobileDisplay mobileDisplay;
    EmailAlert emailAlert(100.0);

    ticker.attach(&mobileDisplay);
    ticker.attach(&emailAlert);

    std::cout << "Price update 1:" << std::endl;
    ticker.setPrice(95.0);

    std::cout << "Price update 2:" << std::endl;
    ticker.setPrice(102.5);

    std::cout << "Detaching EmailAlert." << std::endl;
    ticker.detach(&emailAlert);

    std::cout << "Price update 3:" << std::endl;
    ticker.setPrice(110.0);

    return 0;
}

File: v1/LLD/Design-Patterns/Observer/code/observer.js
/**
 * Observer Pattern — a StockTicker (Subject) notifies registered Observers on price change.
 * Run: node observer.js
 */

class StockTicker {
  constructor(symbol) {
    this.symbol = symbol;
    this.price = 0;
    this.observers = [];
  }

  attach(observer) {
    this.observers.push(observer);
  }

  detach(observer) {
    this.observers = this.observers.filter(o => o !== observer);
  }

  setPrice(newPrice) {
    this.price = newPrice;
    this.notifyObservers();
  }

  notifyObservers() {
    this.observers.forEach(observer => observer.update(this.symbol, this.price));
  }
}

class MobileDisplay {
  update(symbol, price) {
    console.log(`[MobileDisplay] ${symbol} is now $${price}`);
  }
}

class EmailAlert {
  constructor(threshold) {
    this.threshold = threshold;
  }

  update(symbol, price) {
    if (price >= this.threshold) {
      console.log(`[EmailAlert] ${symbol} crossed $${this.threshold} — now $${price}`);
    }
  }
}

const ticker = new StockTicker("ACME");

const mobileDisplay = new MobileDisplay();
const emailAlert = new EmailAlert(100.0);

ticker.attach(mobileDisplay);
ticker.attach(emailAlert);

console.log("Price update 1:");
ticker.setPrice(95.0);

console.log("Price update 2:");
ticker.setPrice(102.5);

console.log("Detaching EmailAlert.");
ticker.detach(emailAlert);

console.log("Price update 3:");
ticker.setPrice(110.0);

module.exports = { StockTicker, MobileDisplay, EmailAlert };
